### PR TITLE
Adding Direct Method APIs to OPC Publisher.

### PR DESCRIPTION
**Pull request by karok2m**
This PR adds support of IoT Edge direct methods in OPC Publisher and adds several accompanying features.

The following IoT Edge direct methods have been added along with tests for them:

* `PublishNodes_V1`: to add a set of nodes to be published for a specific `DataSetWriter`.
* `UnpublishNodes_V1`: to remove nodes from a previously configured `DataSetWriter`.
* `UnpublishAllNodes_V1`: to remove all the nodes from a previously configured `DataSetWriter`.
* `GetConfiguredEndpoints_V1`: to get list of all configured `DataSetWriters`.
* `GetConfiguredNodesOnEndpoint_V1`: to get list of configured nodes for previously configured `DataSetWriter`.
* `GetDiagnosticInfo_V1`: to get diagnostics information for all actively processed `DataSetWriters`.
* `AddOrUpdateEndpoints_V1`:  to perform a complete `published_nodes.json` update or update multiple `DataSetWriters` at once.

Other major changes introduced by this PR

* Added configurable default value for maximum nodes per dataset/subscription. (#1498)
* Added support for incremental publisher reconfiguration. (#1545)

List of all changes:

* Added infrastructure support for device methods in publisher. (#1494)
* Added configurable default value for maximum nodes per dataset/subscription. (#1498)
* Added stress test with 100 endpoints. (#1507)
* Added unit tests for direct method api & bugfixes. (#1502)
* Added direct method E2E `PublishNodes_V1` and `UnpublishNodes_V1` test. (#1508)
* Implemented response payload for `PublishNodes_V1` direct method. (#1500)
* Added response for `UnpublishNodes_V1` direct method. (#1515)
* Added persistency to published nodes file. (#1509)
* Added note about calling `PersistPublishedNodes()` method from a locked block. (#1524)
* Added `GetConfiguredNodesOnEndpoint_V1` direct method. (#1522)
* Fixed data set detection bug in GetConfiguredNodesOnEndpointAsync(). (#1532)
* Added direct method `GetConfiguredEndpoints_V1`. (#1520)
* Added `GetConfiguredNodesOnEndpoint_V1` direct method to E2E test. (#1533)
* Added documentation of OPC Publisher direct methods. (#1538)
* Added E2E tests for `GetConfiguredNodesOnEndpoint_V1` direct method. (#1536)
* Updated publisher.md with `GetConfiguredNodesOnEndpoint_V1` direct method doc. (#1541)
* Updated `GetConfiguredEndpoints_V1` in docs. (#1542)
* Added implementation of `AddOrUpdateEndpoints_V1` direct method call. (#1539)
* Implemented `UnpublishAllNodes_V1` direct method. (#1543)
* Implemented `GetDiagnosticInfo_V1` direct method. (#1544)
* Added support for incremental publisher reconfiguration. (#1545)
* `PublishNodes_V1`/`UnpublishNodes_V1`/`AddOrUpdateEndpoints_V1`/`UnpublishAllNodes_V1` nodes empty response. (#1554)

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs [hunk 1]
```diff
@@ -4,91 +4,88 @@
...
-    using System.Collections.Concurrent;
-    using System.Linq;
-    using Microsoft.Azure.IIoT.OpcUa.Publisher.Models;
-    using Microsoft.Azure.IIoT.Exceptions;
 
     /// <summary>
     /// Job orchestrator the represents the legacy publishednodes.json with legacy command line arguments as job.
     /// </summary>
-    public class LegacyJobOrchestrator : IJobOrchestrator {
+    public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices, IDisposable {
```
Comment (marcschier) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs]: Eventually we should call this "StandaloneJobOrchestrator" or "LocalJobOrchestrator"

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs [hunk 2]
```diff
@@ -141,173 +162,1014 @@ public class LegacyJobOrchestrator : IJobOrchestrator {
...
-        private static string GetChecksum(string file) {
-            using (var stream = File.OpenRead(file)) {
-                var sha = new SHA256Managed();
-                var checksum = sha.ComputeHash(stream);
-                return BitConverter.ToString(checksum).Replace("-", string.Empty);
+        private static string GetChecksum(string content) {
+            if (string.IsNullOrEmpty(content)) {
+                return null;
+            }
+            var sha = new SHA256Managed();
```
Comment (marcschier) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs]: Check if this is disposable and use using

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs (hunk 1 above)
Comment (karok2m) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs]: Renamed everything legacy to standalone.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs (hunk 2 above)
Comment (karok2m) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs]: Fixed to use `using`.

File: api/src/Microsoft.Azure.IIoT.OpcUa.Api.Publisher/src/Models/PublishedNodeApiModel.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            EmitDefaultValue = false)]
+        public string DataSetFieldId { get; set; }
+
+        /// <summary> Display name </summary>
+        [DataMember(Name = "displayName", Order = 7,
+            EmitDefaultValue = false)]
+        public string DisplayName { get; set; }
+
+        /// <summary>
+        /// Heartbeat interval as TimeSpan.
```
Comment (cristipogacean) [inline api/src/Microsoft.Azure.IIoT.OpcUa.Api.Publisher/src/Models/PublishedNodeApiModel.cs:72]: >Heartbeat interval as TimeSpan [](http://example.com/codeflow?start=12&length=30)

in seconds

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs [hunk 4]
```diff
@@ -28,6 +28,7 @@ public class MethodCallStatusException : MethodCallException {
         /// <param name="errorMessage"></param>
         public MethodCallStatusException(int result, string errorMessage = null) :
             this("{}", result, errorMessage) {
+            ResponsePayload = $"\"{errorMessage}\"";
```
Comment (cristipogacean) [inline common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs:31]: >ResponsePayload [](http://example.com/codeflow?start=12&length=15)

change to json encodable object

File: common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs [hunk 5]
```diff
@@ -407,47 +459,49 @@ private class JobProcess : IDisposable {
...
+                                JobHash = Job.GetHashSafe(),
+                                Status = Job.LifetimeData.Status,
+                                ProcessMode = _currentJobProcessInstruction.ProcessMode.Value,
+                                State = await _currentProcessingEngine.GetCurrentJobState().ConfigureAwait(false)
+                            }
+                        }, GetProcessDiagnosticInfo(), ct)
+                    .ConfigureAwait(false);
 
                 // Process instructions
                 switch (result.HeartbeatInstruction) {
```
Comment (cristipogacean) [inline common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs:480]: >switch [](http://example.com/codeflow?start=16&length=6)

add a default

**Review by cristipogacean — DISMISSED** [review]
:shipit:

File: api/src/Microsoft.Azure.IIoT.OpcUa.Api.Publisher/src/Models/PublishedNodeApiModel.cs (hunk 3 above)
Comment (cristipogacean) [reply api/src/Microsoft.Azure.IIoT.OpcUa.Api.Publisher/src/Models/PublishedNodeApiModel.cs:72]: fix in a next commit

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs (hunk 4 above)
Comment (cristipogacean) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs:31]: fix in a next commit

File: common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs (hunk 5 above)
Comment (cristipogacean) [reply common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs:480]: fix in a next commit

**Review by cristipogacean — DISMISSED** [review]
:shipit:

**Review by cristipogacean — DISMISSED** [review]

**Review by cristipogacean — APPROVED** [review]

**Pull request merged**